Add AppDefaults

jiaqiluo opened the pull request:
Issue: https://github.com/rancher/rancher/issues/36827

# Problem:
We use channelserver to host rke2/k3s version information in rancher/rancher, and we need the ability to set different default versions for different rancher versions.  
This ability should be general enough to support any app and not tied to Rancher. 

# Solution:
This PR adds the support for setting app defaults in the config file. 
An example of the value looks like this 
```
appDefaults:
    - appName: rancher
      defaults:
        - appVersion: '> 2.6.2-0 < 2.6.3-0'
          defaultVersion: '1.20.x'
        - appVersion: '> 2.6.2-0 < 2.6.4-0'
          defaultVersion: '1.21.x'
        - appVersion: '> 2.6.4-0'
          defaultVersion: '1.22.x'
    - appName: anotherApp
      defaults:
        - appVersion: '1.2.x'
          defaultVersion: '1.19.5'
```

# Problem:
In the case where the 404 NotFound is returned from the HTTP GET request, channelserver cannot parse the response's data and is stuck in returning errors. 

# Solution:
This PR adds checks for the response's status code and continues to parse the data only if it is 200.


# Problem:
The default value for the flag `path-prefix` is not applied. 

# Solution:
It is a bug in the urfave/cli module: https://github.com/urfave/cli/issues/1121
This PR bumps github.com/urfave/cli/v2 to v2.4.0 to get the fix 

# Others Improvements:
- update error messages to be clear 
- update the readme and default config file

thedadams commented on pkg/config/get.go:
```diff
@@ -146,3 +149,37 @@ func GetGHReleases(ctx context.Context, client *github.Client, owner, repo strin
+		if err := yaml.Unmarshal(content, &data); err != nil {
+			return nil, err
+		}
+		data, _ = data[subKey].(map[string]interface{})
```
Let's check this potential casting error, just in case.

thedadams commented on pkg/config/get.go:
```diff
@@ -45,6 +45,9 @@ func get(ctx context.Context, url Source) ([]byte, error) {
 	if err != nil {
 		return nil, err
 	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("status %v", resp.Status)
+	}
```
This needs to come after the `defer resp.Body.Close()`

thedadams commented on main.go:
```diff
@@ -56,6 +57,12 @@ func main() {
 		},
+		&cli.StringFlag{
+			Name:        "app-name",
+			Usage:       "the app for which the app default versions are for",
```
```suggestion
			Usage:       "the app for which to retrieve default versions",
```

thedadams reviewed (changes requested):
A few minor things.

jiaqiluo commented:
@thedadams  Please check[ the latest commit](https://github.com/rancher/channelserver/pull/17/commits/51f595abf5f59ec0198384b5fb0b519028aa34a4) that addresses your comments.

thedadams commented on pkg/config/get.go:
```diff
@@ -146,3 +150,40 @@ func GetGHReleases(ctx context.Context, client *github.Client, owner, repo strin
+		}
+		data, ok := data[subKey].(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("failed to do the type assertion")
```
```suggestion
			return nil, fmt.Errorf("content at key %s expected to be map[string]interface{}, found %T instead", subKey, data[subKey])
```

thedadams commented on pkg/config/get.go:
```diff
@@ -146,3 +150,40 @@ func GetGHReleases(ctx context.Context, client *github.Client, owner, repo strin
+		if err := yaml.Unmarshal(content, &data); err != nil {
+			return nil, err
+		}
+		data, ok := data[subKey].(map[string]interface{})
```
Avoid variable shadowing so we can make the error message more descriptive.
```suggestion
		subData, ok := data[subKey].(map[string]interface{})
```

thedadams commented on pkg/config/get.go:
```diff
@@ -146,3 +150,40 @@ func GetGHReleases(ctx context.Context, client *github.Client, owner, repo strin
+		if !ok {
+			return nil, fmt.Errorf("failed to do the type assertion")
+		}
+		if err := convert.ToObj(data, &allConfigs); err != nil {
```
```suggestion
		if err := convert.ToObj(subData, &allConfigs); err != nil {
```

jiaqiluo commented:
@thedadams  Please check the latest commit: https://github.com/rancher/channelserver/pull/17/commits/41efe6ca59f80683dd0e006597a478d16de0ee51

thedadams commented on pkg/config/get.go:
```diff
@@ -146,3 +150,40 @@ func GetGHReleases(ctx context.Context, client *github.Client, owner, repo strin
+		}
+		subData, ok := data[subKey].(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("content at key %s expected to be map[string]interface{}, found %T instead", subKey, subData[subKey])
```
```suggestion
			return nil, fmt.Errorf("content at key %s expected to be map[string]interface{}, found %T instead", subKey, data[subKey])
```

jiaqiluo committed:
This PR adds a new data structure AppDefault which maps an app's version range to a release range and can be used to set the default release version(s) per app version(s).
This PR improves the handling of HTTP errors, especially the 404 not found, from the source URL.
This PR also bumps github.com/urfave/cli/v2 to v2.4.0 to fix the issue where the default value is not set for string slice flags.
